src/main/java/com/gregtechceu/gtceu/common/item/ColorSprayBehaviour.java

```diff
@@ -194,17 +195,52 @@ public InteractionResult useOn(UseOnContext context) {
         var facing = context.getClickedFace();
         var pos = context.getClickedPos();
         var stack = context.getItemInHand();
+        var block = level.getBlockState(pos).getBlock();
+
+        int maxBlocksToRecolor = Math.max(1, player != null && player.isCrouching() ? ConfigHolder.INSTANCE.tools.sprayCanChainLength : 1);
+
         if (player != null) {
-            if (!tryPaintBlock(player, level, pos, facing)) {
-                return InteractionResult.PASS;
+            for (int i = 0; i < maxBlocksToRecolor; i++) {
+
+                // Stop once we hit another block to prevent accidentally recoloring stuff
+                if (level.getBlockState(pos).getBlock() != block) {
+                    break;
+                }
+
+                if (!tryPaintBlock(player, level, pos, facing)) {
+                    return InteractionResult.PASS;
+                }
+
+                if (!useItemDurability(player, context.getHand(), stack, empty.get())) {
+                    break;
+                }
```

Comment:
maybe this should before the `tryPaintBlock` call so that the player doesn't get a free spray?

Reply:
We only ever get here when the spray can has atleast 1 use left so it's not free. And once the spray can is emptied, we break the for loop in line 211. On the other side if we'd move it before the `tryPaintBlock` a use is consumed even if the block is uncolorable.
I tested in dev env with an almost empty spray can and it only colored as many blocks as there were uses left.

Reply:
fair enough.